### PR TITLE
drivers: mutex for clock and regulator valid during PM suspend/resume sequences

### PR DESCRIPTION
**Pull request by etienne-lms**
New proposal for the change posted in https://github.com/OP-TEE/optee_os/pull/6724.

This P-R propose a (somewhat) generic implementation for the "PM aware" mutex and apply it the the clock framework (to replace current spinning lock) and in the regulator framework (which already implements the sequence that was seen to weak during https://github.com/OP-TEE/optee_os/pull/6724 review.

### CODE REVIEW

File: core/include/kernel/mutex.h [hunk 1]
```diff
@@ -76,6 +76,27 @@ void mutex_unlock_recursive(struct recursive_mutex *m);
...
+ * non-secure world is not operational, for example in fastcall SMC entries
+ * of the PSCI services. In such case we cannot take a mutex and we expect
+ * the mutex is unlocked. Additionally a spinning lock is attempted to be
+ * locked to check the resource access consistency.
+ */
+struct mutex_pm_aware {
+	struct mutex mutex;	/* access protection in thread context */
+	unsigned int lock;	/* access consistency in PM context */
+};
+
```
Comment (jenswi-linaro) [inline core/include/kernel/mutex.h:94]: Wouldn't a `MUTEX_PM_AWARE_INITIALIZER` make sense too?

File: core/include/kernel/mutex.h [hunk 2]
```diff
@@ -76,6 +76,27 @@ void mutex_unlock_recursive(struct recursive_mutex *m);
...
+/*
+ * struct mutex_pm_aware - Mutex usable in PM atomic sequence
+ *
+ * Some resources need a mutex protection for runtime operations but are
+ * also accessed during specific system power transition (PM power off,
+ * suspend and resume) that operate in atomic execution environment where
+ * non-secure world is not operational, for example in fastcall SMC entries
+ * of the PSCI services. In such case we cannot take a mutex and we expect
+ * the mutex is unlocked. Additionally a spinning lock is attempted to be
+ * locked to check the resource access consistency.
```
Comment (jenswi-linaro) [inline core/include/kernel/mutex.h:88]: Please mention that `mutex_pm_aware_lock()` is expected to panic if it detects contention with an atomic context, regardless if called from a threaded or atomic context, but that contention between two threaded contexts is OK.

File: core/include/drivers/regulator.h [hunk 3]
```diff
@@ -87,7 +87,7 @@ struct regulator_voltages_desc {
  * @max_uv: Max possible voltage level in microvolt (uV)
  * @flags: REGULATOR_* property flags
  * @refcount: Regulator enable request reference counter
- * @lock: Mutex for concurrent access protection
+ * @mu: Mutex for concurrent access protection with PM context support
```
Comment (jenswi-linaro) [inline core/include/drivers/regulator.h:90]: s/mu/mutex/

File: core/include/kernel/mutex.h (hunk 1 above)
Comment (etienne-lms) [reply core/include/kernel/mutex.h:94]: Sorry, it's not for that reason. But the conflict is there. I'll try to find a way to define it and will tell you.

**Comment by etienne-lms** [comment]
Comments addressed.

File: core/include/kernel/mutex_pm_aware.h [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+ * struct mutex_pm_aware - Mutex usable in PM atomic sequence
+ *
+ * Some resources need a mutex protection for runtime operations but are
+ * also accessed during specific system power transition (PM power off,
+ * suspend and resume) that operate in atomic execution environment where
+ * non-secure world is not operational, for example in fastcall SMC entries
+ * of the PSCI services. In such case we cannot take a mutex and we expect
+ * the mutex is unlocked. Additionally a spinning lock is attempted to be
+ * locked to check the resource access consistency. In case of contention
+ * between thread and non-thread access to the resource, core will
```
Comment (jenswi-linaro) [inline core/include/kernel/mutex_pm_aware.h:21]: Contention of non-threaded access always results in a panic. Only contention between threaded accesses will be handled gracefully. This description needs to be rephrased to make that clear.

Please split the comment into two paragraphs, one with the background and one short that describes the intended behavior.

File: core/kernel/mutex.c [hunk 5]
```diff
@@ -342,6 +343,40 @@ unsigned int mutex_get_recursive_lock_depth(struct recursive_mutex *m)
 	return refcount_val(&m->lock_depth);
 }
 
+void mutex_pm_aware_init(struct mutex_pm_aware *m)
+{
+	mutex_init(&m->mutex);
```
Comment (jenswi-linaro) [inline core/kernel/mutex.c:348]: Please use `MUTEX_PM_AWARE_INITIALIZER`.

File: core/include/kernel/mutex_pm_aware.h (hunk 4 above)
Comment (etienne-lms) [reply core/include/kernel/mutex_pm_aware.h:21]: Ok, I understand my phrasing is confusing. I'll do.

File: core/kernel/mutex.c (hunk 5 above)
Comment (etienne-lms) [reply core/kernel/mutex.c:348]: Right.

**Comment by etienne-lms** [comment]
comment addressed

File: core/include/kernel/mutex_pm_aware.h [hunk 6]
```diff
@@ -17,9 +17,13 @@
...
- * locked to check the resource access consistency. In case of contention
- * between thread and non-thread access to the resource, core will
- * intentionally panic.
+ * locked to check the resource access consistency.
+ *
+ * In case a thread request the mutex while a non-thread access still holds
+ * it, core will intentionally panic since it is not expected.
+ *
+ * In case a non-thread access request the mutex while it is held by a thread,
+ * core will also panic since it is not an expected situation.
```
Comment (jenswi-linaro) [inline core/include/kernel/mutex_pm_aware.h:26]: Please add the third case with non-threaded access while another core holds the spinlock.

Comment (etienne-lms) [reply core/include/kernel/mutex_pm_aware.h]: Ok. Comments addressed.

**Review by jenswi-linaro — COMMENTED** [review]
`Reviewed-by: Jens Wiklander <jens.wiklander@linaro.org>`

File: core/drivers/clk/clk.c [hunk 7]
```diff
@@ -7,19 +7,38 @@
...
 static SLIST_HEAD(, clk) clock_list = SLIST_HEAD_INITIALIZER(clock_list);
 #endif
 
+/*
+ * Clock operation may occur at runtime and during specific
+ * system power transition (PM power off, suspend and resume)
+ * that operate for example in fastcall SMC entries of the PSCI
+ * services where non-secure world is not operational. In these
+ * cases we cannot take a mutex and will expect the mutex is
+ * unlocked.
```
Comment (etienne-lms) [inline core/drivers/clk/clk.c:30]: Maybe should I simplify this description? (and ditto for the equiv inline comment in _regulator.c_)?
```
/*
 * Clock operation may occur at runtime and during specific
 * system power transition (PM power off, suspend and resume)
 * hence use a PM aware mutex.
 */

**Review by etienne-lms — COMMENTED** [review]
Fixup commits squashed and review tag applied.

**Comment by etienne-lms** [comment]
I've removed the [RFC] tag from the P-R header line.

File: core/drivers/clk/clk.c (hunk 7 above)
Comment (jenswi-linaro) [reply core/drivers/clk/clk.c:30]: Sure, or even better refer to the description of the `struct mutex_pm_aware`.

Comment (etienne-lms) [reply core/drivers/clk/clk.c]: I've simplified the description.
clk.c: `/* Global clock tree access protection complying the power state transitions */`
regualtor.c: `/* Access protection mutex complying the power state transitions context */`

**Comment by etienne-lms** [comment]
Comments addressed.

**Comment by etienne-lms** [comment]
Thanks

**Pull request merged**